### PR TITLE
[DS-4947] Add negative user sampling to evaluation

### PR DESCRIPTION
**Pull request by emmafgrimes**
## Describe the big picture - why are we doing this?
This is the (hopefully) final PR in adding the ability to negative sample users in `collie` models! Specifically, this is the PR that adds the ability to `metrics.py`. I also caught a couple other places in `utils.py` and `cross_validation.py` that needed updating as well. Please take your time reviewing this PR, it was the most difficult of the 3 for me. 
 
## Any additional details to clarify code in the PR?
The metrics code can be hard to follow at times, so as always, I am open to feedback in enhancing the readability with any refactoring.

## How is this tested?
On an EC2 instance all the tests passed. We have one final ticket for this work where I will rerun all the tutorial notebooks with negative user sampling. 

## Pull Request Checklist
 - [x] Pull request includes a description of why we are doing this
 - [x] ``CHANGELOG`` has been updated
 - [x] Version in ``_version.py`` has been updated
 - [x] All tests in the ``tests`` folder pass with a local build
 - [ ] ``README`` has been updated (if applicable)
 - [ ] Documentation in ``docs`` has been updated (if applicable)
 - [ ] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)
 - [x] Docker image can be built using ``docker build -t collie .``

### CODE REVIEW

File: collie/cross_validation.py [hunk 1]
```diff
@@ -237,42 +239,53 @@ def _stratified_split(interactions: BaseInteractions,
...
-    users = interactions.mat.row
-    unique_users = set(users)
-
+    try:
+        if interactions.negative_sample_type == 'item':
+            users_or_items = interactions.mat.row
+        elif interactions.negative_sample_type == 'user':
+            users_or_items = interactions.mat.col
+    except AttributeError:
+        users_or_items = interactions.mat.row
```
Comment (nathancooperjones) [inline collie/cross_validation.py]: I think we can simplify this logic a bit more: 

```suggestion
    users_or_items = interactions.mat.row
    
    if getattr(interactions, 'negative_sample_type') == 'user':
        users_or_items = interactions.mat.col
```

File: collie/cross_validation.py [hunk 2]
```diff
@@ -282,10 +295,19 @@ def _stratified_split(interactions: BaseInteractions,
...
                                       test_p: float,
                                       seed: int,
                                       force_split: bool) -> np.array:
+    try:
+        if interactions.negative_sample_type == 'item':
+            opposite_type = 'user'
+        elif interactions.negative_sample_type == 'user':
+            opposite_type = 'item'
+    except AttributeError:
+        opposite_type = 'user'
```
Comment (nathancooperjones) [inline collie/cross_validation.py]: Could we try just passing in `opposite_type` to each worker in the parent function above so we don't have to re-compute this each time.

File: collie/metrics.py [hunk 3]
```diff
@@ -49,129 +53,159 @@ def _get_user_item_pairs(user_ids: Union[np.array, torch.tensor],
...
 
     """
-    # Added because sometimes we call this function with ``n_items`` as ``np.int64`` type which
-    # breaks ``repeat_interleave``.
-    if isinstance(n_items, np.int64):
-        n_items = n_items.item()
+    try:
+        negative_sample_type = model.train_loader.negative_sample_type
+    except AttributeError:  # explicit models do not have ``negative_sample_type``
+        negative_sample_type = 'item'
```
Comment (nathancooperjones) [inline collie/metrics.py]: Same comment here - can we just pass in `negative_sample_type` here to avoid re-computing this each time?

File: collie/metrics.py [hunk 4]
```diff
@@ -8,33 +8,37 @@
...
 
     Parameters
     ----------
-    user_ids: np.array or torch.tensor, 1-d
-        Iterable[int] of users to score
-    n_items: int
-        Number of items in the training data
+    model: collie.model.BasePipeline
+        Model that can take a (user_id, item_id) pair as input and return a recommendation score
+    user_or_item_ids: np.array or torch.tensor, 1-d
```
Comment (nathancooperjones) [inline collie/metrics.py]: This is a time when I wish we had a Nicole around. For now, I think this works, but before we think about merging into the main Collie codebase, I think we should try to establish a better naming scheme that makes the code more readable while still understanding that things can refer to either users or items. I hate that we have to put `user_or_item` everywhere now, but I have no better ideas.

File: collie/metrics.py [hunk 5]
```diff
@@ -49,129 +53,159 @@ def _get_user_item_pairs(user_ids: Union[np.array, torch.tensor],
...
+        negative_sample_type = 'item'
 
-    users = torch.tensor(
-        user_ids,
+    # Added because sometimes we call this function with ``n_items_or_users`` as ``np.int64``
+    # type which breaks ``repeat_interleave``.
+    if isinstance(n_items_or_users, np.int64):
+        n_items_or_users = n_items_or_users.item()
+
+    aabb_tensor = torch.tensor(
```
Comment (nathancooperjones) [inline collie/metrics.py]: What do the names `aabb` and `cdcd` mean?

File: collie/metrics.py [hunk 6]
```diff
@@ -8,33 +8,37 @@
...
         Device to store tensors on
 
     Returns
     -------
     users: torch.tensor, 1-d
-        Tensor with ``n_items`` copies of each user ID
+        Tensor with ``n_items_or_users`` copies of each user ID
     items: torch.tensor, 1-d
-        Tensor with ``len(user_ids)`` copies of each item ID
+        Tensor with ``len(user_or_item_ids)`` copies of each item ID
```
Comment (nathancooperjones) [inline collie/metrics.py]: We also need to update the `Example` section of the docs below this.

File: collie/metrics.py [hunk 7]
```diff
@@ -49,129 +53,159 @@ def _get_user_item_pairs(user_ids: Union[np.array, torch.tensor],
...
+            device=device,
+        )
+    elif negative_sample_type == 'user':
+        return torch.tensor(
+            (targets[np.array(preds.detach().cpu()), user_or_item_ids[:, None]] > 0)
+            .astype('double')
+            .toarray(),
+            requires_grad=False,
+            device=device,
+        )
```
Comment (nathancooperjones) [inline collie/metrics.py]: We can reduce repeated code here a bit more: 

```suggestion
    if negative_sample_type == 'item':
        labels = targets[user_or_item_ids[:, None], np.array(preds.detach().cpu())]
    elif negative_sample_type == 'user':
        labels = targets[np.array(preds.detach().cpu()), user_or_item_ids[:, None]]

    return torch.tensor(
        (labels > 0)
        .astype('double')
        .toarray(),
        requires_grad=False,
        device=device,
    )
```

File: collie/metrics.py [hunk 8]
```diff
@@ -182,12 +216,18 @@ def mapk(targets: csr_matrix,
...
+    if negative_sample_type == 'item':
+        denominator = torch.min(
+            torch.tensor(k, device=device, dtype=torch.int).repeat(len(user_or_item_ids)),
+            torch.tensor(targets[user_or_item_ids].getnnz(axis=1), device=device)
+        )
+    elif negative_sample_type == 'user':
+        denominator = torch.min(
+            torch.tensor(k, device=device, dtype=torch.int).repeat(len(user_or_item_ids)),
+            torch.tensor(targets[:, user_or_item_ids].getnnz(axis=0), device=device)
+        )
```
Comment (nathancooperjones) [inline collie/metrics.py]: I think we can also refactor this a bit more to reduce repeated code.

File: collie/metrics.py [hunk 9]
```diff
@@ -196,37 +236,50 @@ def mapk(targets: csr_matrix,
...
-
-    # weighting each 0/1 by position so that topk returns index of *first* postive result
-    position_weight = 1.0/(
-        torch.arange(1, targets.shape[1] + 1, device=preds.device)
-        .repeat(len(user_ids), 1)
-        .float()
-    )
+    device = preds.device
+    predictions = preds.topk(preds.shape[1], dim=1).indices
+    labeled = _get_labels(targets, user_or_item_ids, predictions, device, negative_sample_type)
```
Comment (nathancooperjones) [inline collie/metrics.py]: [Optional] I realize the original code wasn't written this way, but my new coding philosophy is to be explicit with keyword arguments as much as possible. I think it's a lot clearer (and less error-prone) to write function calls like this: 

```python
foo(a=a, b=b, c=c, d=d)
```

instead of 

```python
foo(a, b, c, d)
```

If you have energy and want to, I think it would be nice to do that for all of our internal function calls in this file. If not, feel free to ignore.

File: collie/metrics.py [hunk 10]
```diff
@@ -196,37 +236,50 @@ def mapk(targets: csr_matrix,
...
+            .repeat(len(user_or_item_ids), 1)
+            .float()
+        )
+    elif negative_sample_type == 'user':
+        # weighting each 0/1 by position so that topk returns index of *first* postive result
+        position_weight = 1.0/(
+            torch.arange(1, targets.shape[0] + 1, device=device)
+            .repeat(len(user_or_item_ids), 1)
+            .float()
+        )
```
Comment (nathancooperjones) [inline collie/metrics.py]: Ditto here with less repeated code.

File: collie/metrics.py [hunk 11]
```diff
@@ -238,42 +291,59 @@ def mrr(targets: csr_matrix,
...
+            target_tensor = torch.tensor(
+                targets[:, item_id].toarray(),
+                device=preds.device,
+                dtype=torch.long
+            ).view(-1)
+            # many models' ``preds`` may be unbounded if a final activation layer is not applied
+            # we have to normalize ``preds`` here to avoid a ``ValueError`` stating that ``preds``
+            # should be probabilities, but values were detected outside of [0,1] range
+            auc = auroc(torch.sigmoid(preds[i, :]), target=target_tensor, pos_label=1)
+            agg += auc
```
Comment (nathancooperjones) [inline collie/metrics.py]: Ditto here.

File: tests/fixtures/metrics_fixtures.py [hunk 12]
```diff
@@ -23,6 +23,16 @@ def test_implicit_predicted_scores():
...
 
 
+@pytest.fixture()
+def test_implicit_predicted_scores_negative_sample_user():
+    return torch.tensor([
+        [0.9, 0.1, -1],
+        [0.4, 1.2, -2],
+        [0.5, 1.7, -3],
+        [0.7, 0.12, 4],
+    ])
```
Comment (nathancooperjones) [inline tests/fixtures/metrics_fixtures.py]: It looks like this is just `test_implicit_predicted_scores` but transposed. If that's intentional, can we just use that logic to make this fixture, rather than have it fully typed out?

File: tests/test_metrics.py [hunk 13]
```diff
@@ -24,38 +24,69 @@ def get_model_scores(user, item, scores):
...
                                                       device=device)
 
     assert torch.equal(actual_users, expected_users)
     assert torch.equal(actual_items, expected_items)
 
 
+@mock.patch('collie.model.MatrixFactorizationModel')
 @pytest.mark.parametrize('n_items_type', ['int', 'np.int64'])
-def test_get_user_item_pairs_unordered(device, n_items_type):
+def test_get_user_item_pairs_unordered(model, device, n_items_type):
```
Comment (nathancooperjones) [inline tests/test_metrics.py]: Do we need to test the user-based version of this as well?

File: tests/fixtures/metrics_fixtures.py [hunk 14]
```diff
@@ -59,11 +79,33 @@ def test_implicit_labels():
...
+@pytest.fixture()
+def targets_auc_negative_sample_user():
+    # ``sklearn.metrics.roc_auc_score`` throws an error if an item has interacted
+    # with every user so changing the final item
+    return Interactions(
+        users=[0, 0, 0, 1, 1, 1, 2, 2],
+        items=[0, 1, 2, 1, 2, 3, 0, 3],  # changed final item
+        ratings=[1, 1, 1, 1, 1, 1, 1, 1],
+        check_num_negative_samples_is_valid=False,
+    ).mat.tocsr()
```
Comment (nathancooperjones) [inline tests/fixtures/metrics_fixtures.py]: Why make this a separate fixture and not just update the original `targets` fixture?

File: tests/test_metrics.py [hunk 15]
```diff
@@ -238,6 +365,7 @@ def test_evaluate_in_batches(
     metrics,
     batch_size,
 ):
+    model.train_loader.negative_sample_type = 'item'
```
Comment (nathancooperjones) [inline tests/test_metrics.py]: Should we test `test_evaluate_in_batches` for user-based negative sampling as well?

File: tests/test_utils.py [hunk 16]
```diff
@@ -114,9 +114,26 @@ def test_remove_users_with_fewer_than_n_interactions(df_with_users_interacting_o
...
+    expected = pd.DataFrame(data={
+        'userId': [1, 2, 3, 4, 5, 6, 7, 8, 9, 2],
+        'itemId': [0, 1, 1, 2, 2, 2, 3, 3, 3, 0],
+    })
+
+    actual = remove_users_or_items_with_fewer_than_n_interactions(
+        df=df_with_items_interacting_only_once,
+        min_num_of_interactions=2,
+        user_or_item_col='itemId'
+    )
```
Comment (nathancooperjones) [inline tests/test_utils.py]: I actually think we don't need to add this test, since we just use the column name directly. The above test already tests 100% of the code.

File: CHANGELOG.md [hunk 17]
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).
 
+# [1.3.0] - 2022-3-04
+### Added
+ - ``negative_sample_type`` to ``Interactions`` and ``DataLoader`` classes to have the abilty to negative sample users or items
+ - ``BasePipeline`` inheirits  ``negative_sample_type`` from the ``train_loader``
```
Comment (nathancooperjones) [inline CHANGELOG.md]: Typo. 

```suggestion
 - ``BasePipeline`` inherits  ``negative_sample_type`` from the ``train_loader``
```

File: CHANGELOG.md [hunk 18]
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
...
+ - ``negative_sample_type`` arguements in ``mapk``, ``mrr``, and ``auc`` and the calculations are computed accordingly
```
Comment (nathancooperjones) [inline CHANGELOG.md]: Typo. 

```suggestion
 - ``negative_sample_type`` arguments in ``mapk``, ``mrr``, and ``auc`` and the calculations are computed accordingly
```

Comment (nathancooperjones) [reply CHANGELOG.md]: I'm also not sure we need to add the `and the calculations are computed accordingly` there since I think that is implied.

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
I think this is good work for V1 of CF to deploy at TMP, but for merging back into the main Collie repo, I think we got to talk about the right way to name and document all of this in as little a confusing way as possible.

File: collie/metrics.py (hunk 4 above)
Comment (emmafgrimes) [reply collie/metrics.py]: Per offline discussion, `rows` and `cols` FTW!

File: collie/metrics.py (hunk 5 above)
Comment (emmafgrimes) [reply collie/metrics.py]: It was the pattern of the output tensor, all has changed now!

File: collie/metrics.py (hunk 9 above)
Comment (emmafgrimes) [reply collie/metrics.py]: I think I covered all the cases of this, but please call out any I missed in the next review - happy to edit!

File: collie/cross_validation.py (hunk 1 above)
Comment (emmafgrimes) [reply collie/cross_validation.py]: This raises an `AttributeError` I added a default `if getattr(interactions, 'negative_sample_type', 'item') == 'user':` let me know what you think in the next review.

File: collie/cross_validation.py [hunk 19]
```diff
@@ -239,13 +240,13 @@ def _stratified_split(interactions: BaseInteractions,
...
-        elif interactions.negative_sample_type == 'user':
-            users_or_items = interactions.mat.col
-    except AttributeError:
-        users_or_items = interactions.mat.row
+    users_or_items = interactions.mat.row
+    opposite_type = 'user'
+
+    if getattr(interactions, 'negative_sample_type', 'item') == 'user':
+        users_or_items = interactions.mat.col
+        opposite_type = 'item'
```
Comment (nathancooperjones) [inline collie/cross_validation.py]: Should we do the same `row/col` naming scheme here too? My preference is to eliminate as many `user_or_item` variable names as possible.

File: collie/cross_validation.py [hunk 20]
```diff
@@ -299,7 +302,8 @@ def _stratified_split_parallel_worker(interactions: BaseInteractions,
                                       idxs_to_split: Iterable[Any],
                                       test_p: float,
                                       seed: int,
-                                      force_split: bool) -> np.array:
+                                      force_split: bool,
+                                      opposite_type: Literal['item', 'user'] = 'user') -> np.array:
```
Comment (nathancooperjones) [inline collie/cross_validation.py:307]: Why does this function take `opposite_type` as an input but then recompute it in the line below?

File: collie/metrics.py [hunk 21]
```diff
@@ -15,83 +15,69 @@
...
     Example
     -------
     .. code-block:: python
 
-        >>> user_ids = np.array([10, 11, 12])
-        >>> n_items = 4
-        >>> user, item = _get_user_item_pairs(user_ids: user_ids, n_items: 4, device: 'cpu'):
-        >>> user
+        >>> row_ids = np.array([10, 11, 12])
+        >>> n_cols = 4
```
Comment (nathancooperjones) [inline collie/metrics.py]: Maybe make this example a little clearer with comments: 

```suggestion
        >>> # let's assume an interactions matrix with users as rows and item as columns
        >>> row_ids = np.array([10, 11, 12])  # user IDs 
        >>> n_cols = 4  # number of items
```

File: collie/metrics.py [hunk 22]
```diff
@@ -15,83 +15,69 @@
...
-        end=n_items_or_users,
+        end=n_cols,
         requires_grad=False,
         device=device,
-    ).repeat(len(user_or_item_ids))
-
-    if negative_sample_type == 'item':
-        users = aabb_tensor
-        items = cdcd_tensor
+    ).repeat(len(row_ids))
```
Comment (nathancooperjones) [inline collie/metrics.py:72]: This is so much clearer!

File: collie/metrics.py [hunk 23]
```diff
@@ -15,83 +15,69 @@
...
 
     Parameters
     ----------
-    model: collie.model.BasePipeline
-        Model that can take a (user_id, item_id) pair as input and return a recommendation score
-    user_or_item_ids: np.array or torch.tensor, 1-d
+    row_ids: np.array or torch.tensor, 1-d
         Iterable[int] of users or items to score
-    n_items_or_users: int
+    n_cols: int
```
Comment (nathancooperjones) [inline collie/metrics.py:28]: Should we also change the description of this variable on the next line to not say `items or users` and just be consistent with `row/col`?

Comment (nathancooperjones) [reply collie/metrics.py:28]: Ditto to a lot of docstrings in this file.

File: tests/test_metrics.py [hunk 24]
```diff
@@ -116,98 +89,53 @@ def test_get_preds_implicit_negative_sample_user(
     n_items, n_users = test_implicit_predicted_scores_negative_sample_user.shape
     item_ids = np.arange(n_items)
     model.train_loader.negative_sample_type = 'user'
-    model.return_value = test_implicit_predicted_scores_negative_sample_user.view(-1)
-    actual_preds = get_preds(model=model,
-                             user_or_item_ids=item_ids,
-                             n_items_or_users=n_users,
-                             device=device)
+    model.return_value = test_implicit_predicted_scores_negative_sample_user.reshape(12)
```
Comment (nathancooperjones) [inline tests/test_metrics.py]: Why `12` here? Is there a way to make this more dynamic so in case we change the fixture in the future, the test doesn't have to also be changed?

File: collie/cross_validation.py [hunk 25]
```diff
@@ -145,14 +146,16 @@ def stratified_split(interactions: BaseInteractions,
                      force_split: bool = False) -> Tuple[BaseInteractions, ...]:
     """
     Split an ``Interactions`` instance into train, validate, and test datasets in a stratified
-    manner such that each user appears at least once in each of the datasets.
+    manner such that each user or item appears at least once in each of the datasets. Whether
+    the split is on user of item is determined by the input ``Interactions`` instance.
 
-    This split guarantees that every user will be represented in the training, validation, and
-    testing datasets given they appear in ``interactions`` at least three times. If ``val_p ==
+    This split guarantees that every user or item will be represented in the training, validation,
```
Comment (nathancooperjones) [inline collie/cross_validation.py:153]: I think part of the `row/col` discussion was to make this consistent across all Collie files where we use `user or item`. What do you think about applying that to other files, like this one?

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
This looks _sooo_ much better, nice work! I think this initial work to move towards `row/col` is great, and I'm hoping we can do this in more files/functions. 

Also, I think we should hold off on merging until we address why the tests are taking so long to run.

File: collie/cross_validation.py (hunk 20 above)
Comment (emmafgrimes) [reply collie/cross_validation.py:307]: OOPS!

File: collie/metrics.py (hunk 23 above)
Comment (emmafgrimes) [reply collie/metrics.py:28]: Oh yes, apologies I meant to do that before submitting!

File: tests/test_metrics.py (hunk 24 above)
Comment (emmafgrimes) [reply tests/test_metrics.py]: Unfortunately, `.view(-1)` doesn't work on the transposed tensor like in the test above. I can replace `12` with `torch.numel` tho:
```suggestion
    model.return_value = test_implicit_predicted_scores_negative_sample_user.reshape(
      torch.numel(test_implicit_predicted_scores_negative_sample_user)
    )
```

File: collie/cross_validation.py (hunk 25 above)
Comment (emmafgrimes) [reply collie/cross_validation.py:153]: I am kind of struggling with when to change to `row/col`. For this specific file, I think `user or item` makes more sense since we didn't change the `datasets.py` file and `Interactions` objects are still defined as rows = users and cols = items up to this point. The `row/col` switch only really happens in metrics. Happy to discuss this further.

File: tests/test_metrics.py (hunk 24 above)
Comment (nathancooperjones) [reply tests/test_metrics.py]: Love this!

File: collie/cross_validation.py (hunk 25 above)
Comment (nathancooperjones) [reply collie/cross_validation.py:153]: That makes total sense! Just for the sake of improved clarity, in these docstrings, can we say something along the lines of ```user or item, depending on ``negative_sample_type`` ``` just to make it clear when one applies. 

Maybe even something like: 

```
user or item, depending on ``negative_sample_type`` (user when ``negative_sample_type == 'item'``, item when user when ``negative_sample_type == 'user'``) 
```

although this might be overkill.

File: tests/conftest.py [hunk 26]
```diff
@@ -10,7 +10,7 @@
 from .fixtures.metrics_fixtures import *
 from .fixtures.model_fixtures import *
 from .fixtures.movielens_fixtures import *
-from .fixtures.utils_fixtures import *
+# from .fixtures.utils_fixtures import *
```
Comment (nathancooperjones) [inline tests/conftest.py]: We'll probably want to un-comment this out.

File: tests/test_model.py [hunk 27]
```diff
@@ -59,6 +59,7 @@ def test_CollieTrainer_no_gpu(is_available_mock, untrained_implicit_model, capfd
 
 @mock.patch('torch.cuda.is_available')
 def test_CollieTrainer_on_cpu(is_available_mock, untrained_implicit_model):
+    import pdb; pdb.set_trace()
```
Comment (nathancooperjones) [inline tests/test_model.py]: Same here.

File: tests/conftest.py (hunk 26 above)
Comment (emmafgrimes) [reply tests/conftest.py]: Excuse me I did not request re-review yet ;p

File: collie/cross_validation.py [hunk 28]
```diff
@@ -145,14 +146,17 @@ def stratified_split(interactions: BaseInteractions,
                      force_split: bool = False) -> Tuple[BaseInteractions, ...]:
     """
     Split an ``Interactions`` instance into train, validate, and test datasets in a stratified
-    manner such that each user appears at least once in each of the datasets.
+    manner such that each user or item, depending on ``interactions.negative_sample_type`` (user
+    when ``negative_sample_type == 'item'``, item when ``negative_sample_type == 'user'``) appears
+    at least once in each of the datasets.
```
Comment (nathancooperjones) [inline collie/cross_validation.py:151]: Can we add this language to: 

* `ApproximateNegativeSamplingInteractionsDataLoader` L191 
* `Interactions` L262 
* `BasePipeline.calculate_loss` L591

**Comment by emmafgrimes** [comment]
Keeping an eye on the tests, but this is otherwise ready for re-review.

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
Changes look good! But, the tests are still taking longer to run on some versions of Python. Can we investigate why this is happening before merging in?

**Comment by emmafgrimes** [comment]
> Changes look good! But, the tests are still taking longer to run on some versions of Python. Can we investigate why this is happening before merging in?

Yep! Doing that now...

**Comment by emmafgrimes** [comment]
WOO tests fixed! Thanks for the help @nathancooperjones , back to you!

**Review by nathancooperjones — APPROVED** [review]
LGTM! 🎉 congrats!!
